### PR TITLE
use thread safe time method

### PR DESCRIPTION
**Pull request by hanmenghao1995**

### CODE REVIEW

**Comment by hanmenghao1995** [comment]
#64

File: src/utils.cc [hunk 1]
```diff
@@ -256,7 +256,7 @@ std::string minio::utils::FormatTime(const std::tm* time, const char* format) {
 
 std::tm* minio::utils::Time::ToUTC() {
   std::tm* t = new std::tm;
-  *t = utc_ ? *std::localtime(&tv_.tv_sec) : *std::gmtime(&tv_.tv_sec);
+  utc_ ? localtime_r(&tv_.tv_sec, t) : gmtime_r(&tv_.tv_sec, t);
```
Comment (balamurugana) [inline src/utils.cc:259]: Please apply below diff
```diff
diff --git a/src/baseclient.cc b/src/baseclient.cc
index 184494f..5995733 100644
--- a/src/baseclient.cc
+++ b/src/baseclient.cc
@@ -818,7 +818,6 @@ minio::s3::BaseClient::GetObjectLockConfig(GetObjectLockConfigArgs args) {
   if (!rule) return config;
 
   auto text = rule.node().select_node("DefaultRetention/Mode/text()");
-  RetentionMode* mode = new RetentionMode;
   config.retention_mode = StringToRetentionMode(text.node().value());
 
   if (rule.node().select_node("DefaultRetention/Days")) {
diff --git a/src/utils.cc b/src/utils.cc
index 942da4c..67f5747 100644
--- a/src/utils.cc
+++ b/src/utils.cc
@@ -279,6 +279,7 @@ std::string minio::utils::Time::ToHttpHeaderValue() {
   std::stringstream ss;
   ss << WEEK_DAYS[utc->tm_wday] << ", " << FormatTime(utc, "%d ")
      << MONTHS[utc->tm_mon] << FormatTime(utc, " %Y %H:%M:%S GMT");
+  delete utc;
   return ss.str();
 }
 
```

**Review by balamurugana — COMMENTED** [review]
Please add a test for this case

File: src/utils.cc (hunk 1 above)
Comment (hanmenghao1995) [reply src/utils.cc:259]: > Please add a test for this case

test for this case needs multi threads to verify, and It takes tens of seconds. Is it appropriate as a ut case?

Comment (balamurugana) [reply src/utils.cc:259]: Isn't it possible to reproduce it by having 10 times `ToUTC()` in 100 threads?

**Pull request closed without merging**